### PR TITLE
feat: streaming sink

### PR DESCRIPTION
**Pull request by kohlisid**
Resolves: https://github.com/numaproj/numaflow/issues/2097

### CODE REVIEW

**Review by vigith — APPROVED** [review]

File: pynumaflow/sinker/servicer/sync_servicer.py [hunk 1]
```diff
@@ -1,69 +1,104 @@
...
-        The pascal case function name comes from the proto sink_pb2_grpc.py file.
+        Applies a sink function to datum elements.
         """
-        # if there is an exception, we will mark all the responses as a failure
-        datum_iterator = datum_generator(request_iterator)
+
         try:
-            rspns = self.__sink_handler(datum_iterator)
+            # The first message to be received should be a valid handshake
+            req = next(request_iterator)
```
Comment (ab93) [inline pynumaflow/sinker/servicer/sync_servicer.py:38]: Might be a good idea to handle the `StopIteration` error differently

File: pynumaflow/sinker/servicer/utils.py [hunk 2]
```diff
@@ -1,13 +1,75 @@
...
+    responses (Responses): A list of Response objects containing the results of sink operations.
+
+    Returns:
+    list[sink_pb2.SinkResponse.Result]: A list of SinkResponse.Result objects.
+    """
+    ret = []
+    for rspn in responses:
+        sink_rsp = build_sink_response(rspn)
+        ret.append(sink_rsp)
+    return ret
```
Comment (ab93) [inline pynumaflow/sinker/servicer/utils.py:19]:
```suggestion
    return [build_sink_response(rspn) for rspn in responses]
```

**Review by KeranYang — APPROVED** [review]
Approving with nit comments.

File: pynumaflow/shared/thread_with_return.py
[the remark this answers is not in the document]
Comment (kohlisid) [reply pynumaflow/shared/thread_with_return.py:1]: Yes!

File: pynumaflow/sinker/servicer/sync_servicer.py (hunk 1 above)
Comment (kohlisid) [reply pynumaflow/sinker/servicer/sync_servicer.py:38]: Should be covered by the Base exception right now, but will think about it and update in a follow up.

**Review by ab93 — APPROVED** [review]

**Pull request merged**